### PR TITLE
Support Baichuan

### PR DESCRIPTION
**Pull request by AllentDan**

### CODE REVIEW

**Comment by grimoire** [comment]
Please solve conflict

File: lmdeploy/model.py
Comment (wangruohui) [inline lmdeploy/model.py:1]: How to handle this part with #378?

File: lmdeploy/pytorch_poc/patch/baichuan.py [hunk 1]
```diff
@@ -0,0 +1,327 @@
...
+
+        attn_output = origin_self.o_proj(attn_output)
+
+        if not output_attentions:
+            attn_weights = None
+
+        return attn_output, attn_weights, past_key_value
+
+
+class BaichuanLayer(torch.nn.Module):
```
Comment (wangruohui) [inline lmdeploy/pytorch_poc/patch/baichuan.py:113]: It seems this modification is only for allowing position ids passed into attention to compute q_locs and q_seq_lens for paged attention. Since these two are crucial for paged attention, so I think we can pass these two values consistently through the `context`.to avoid copying/pasting a lot of codes just for changing the argument list.
I have made this work in the chatglm part, https://github.com/InternLM/lmdeploy/pull/360/files#diff-a5adea9960ed4a4cc81eee9f97a1bdb9e8353832cdfc3d56a301a840d35a545bR490. If all (cc @grimoire ) think this is a good idea, I can make a standalone PR for this.

File: lmdeploy/model.py
Comment (AllentDan) [reply lmdeploy/model.py:1]: They are compatible.

File: lmdeploy/pytorch_poc/patch/baichuan.py (hunk 1 above)
Comment (wangruohui) [reply lmdeploy/pytorch_poc/patch/baichuan.py:113]: https://github.com/InternLM/lmdeploy/pull/398

File: lmdeploy/pytorch_poc/patch/baichuan.py [hunk 2]
```diff
@@ -0,0 +1,327 @@
...
+
+        outputs = (hidden_states, )
+
+        if use_cache:
+            outputs += (present_key_value, )
+
+        return outputs
+
+
+class BaichuanModel(nn.Module):
```
Comment (grimoire) [inline lmdeploy/pytorch_poc/patch/baichuan.py:155]: What is the different between this rewrite and the one in llama? Can we register LlamaModel rewrite as BaichuanModel?

File: lmdeploy/pytorch_poc/patch/baichuan.py [hunk 3]
```diff
@@ -0,0 +1,327 @@
...
+        else:
+            return self._continuous_batching_forward(
+                input_ids, attention_mask, position_ids, past_key_values,
+                inputs_embeds, use_cache, output_attentions,
+                output_hidden_states, return_dict)
+
+
+class BaichuanForCausalLM(nn.Module):
+
+    def forward(self,
```
Comment (grimoire) [inline lmdeploy/pytorch_poc/patch/baichuan.py:284]: Is this rewrite for https://github.com/baichuan-inc/Baichuan-7B/blob/6f3ef4633a90c2d8a3e0763d0dec1b8dc11588f5/models/modeling_baichuan.py#L581? I did not see much difference between this one and the origin one.

Comment (AllentDan) [reply lmdeploy/pytorch_poc/patch/baichuan.py:284]: It's for 13B

**Review by grimoire — APPROVED** [review]

**Review by wangruohui — APPROVED** [review]

**Comment by grimoire** [comment]
TODO: alibi mask
